interop/xds/client/client.go

```diff
@@ -0,0 +1,184 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Binary client for xDS interop tests.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	_ "google.golang.org/grpc/balancer/grpclb"
+	"google.golang.org/grpc/grpclog"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/peer"
+)
+
+type statsWatcherKey struct {
+	startID int32
+	endID   int32
+}
+
+type statsWatcher struct {
+	mu            sync.Mutex
+	rpcsByPeer    map[string]int32
+	numFailures   int32
+	remainingRpcs int32
+}
+
+var (
+	numChannels   = flag.Int("num_channels", 1, "Num of channels")
+	printResponse = flag.Bool("print_response", false, "Write RPC response to stdout")
+	qps           = flag.Int("qps", 1, "QPS per channel")
+	rpcTimeout    = flag.Duration("rpc_timeout", 10*time.Second, "Per RPC timeout")
+	server        = flag.String("server", "localhost:8080", "Address of server to connect to")
+	statsPort     = flag.Int("stats_port", 8081, "Port to expose peer distribution stats service")
+
+	mu               sync.Mutex
+	currentRequestId int32 = 0
+	watchers               = make(map[statsWatcherKey]*statsWatcher)
+)
+
+type statsService struct{}
+
+// Wait for the next LoadBalancerStatsRequest.GetNumRpcs to start and complete,
+// and return the distribution of remote peers. This is essentially a clientside
+// LB reporting mechanism that is designed to be queried by an external test
+// driver when verifying that the client is distributing RPCs as expected.
+func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalancerStatsRequest) (*testpb.LoadBalancerStatsResponse, error) {
+	mu.Lock()
+	watcherKey := statsWatcherKey{currentRequestId, currentRequestId + in.GetNumRpcs()}
+	watcher, ok := watchers[watcherKey]
+	if !ok {
+		watcher = &statsWatcher{
+			rpcsByPeer:    make(map[string]int32),
+			numFailures:   0,
+			remainingRpcs: in.GetNumRpcs(),
+		}
+		watchers[watcherKey] = watcher
+	}
+	mu.Unlock()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(in.GetTimeoutSec())*time.Second)
+	defer cancel()
+
+	done := make(chan bool)
+	// Wait until the requested RPCs have all been recorded or timeout occurs.
+	go func() {
+		for {
+			select {
+			default:
```

Comment:
This busy loop is going to cause high cpu usage.

We can make a channel for each watcher, and send something (e.g. the response, and nil if the RPC failed) whenever an RPC that this watcher in interested finishes.

You can do `remainingRpcs--` here after the channel recv.

So this will be something like

```go
select {
  case r := <- watcher.c:
    if r != nil {
      rpcsByPeer[r.GetServerId()] += 1
    } else {
      numFailures += 1
    }
    remainingRPCs--
    if remainingRPCs == 0 {
      done <- true
      return
    }
  case <- ctx.Done():
    done <- false
    return
}
```

Reply:
Done. A little bit better understanding of channels now - I think this allows removing `statsWatcher.mu` altogether.